In [1]:
#==========================================================

import numpy as np # linear algebra
import pandas as pd # data processing

#===========================================================

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score, mean_squared_error

#===========================================================

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor

#==========================================================
from warnings import filterwarnings
filterwarnings("ignore")
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/datasets/nalisha/california-housing-prices-dataset-clean-and-ml/housing.csv.csv


# Data Preprocessing.....

In [2]:
#=======================Load Data==============================

data = pd.read_csv("/kaggle/input/datasets/nalisha/california-housing-prices-dataset-clean-and-ml/housing.csv.csv")
display(data.head())

#===============================================

print("-----Data Shape----\n")
print(data.shape)

#===============================================

print("\n-----Description-----\n")
display(data.describe())

#===============================================

print("\n----Information-----\n")
display(data.info())

#===============================================

print("\n-----Columns Name------\n")
data_1 = pd.DataFrame(data.columns, columns = ["Columns Name"])
display(data_1)

#===============================================

print("\n-----Before Missing Value------\n")
display(data.isnull().sum())

#==============================================

data["total_bedrooms"].fillna(data["total_bedrooms"].median(), inplace=True)
print("\n-----After Missing Value------\n")
display(data.isnull().sum())

#==============================================

print("\n-----Duplicated Value-----\n")
display(data.duplicated().sum())

#==============================================

print("\n------nunique Value-------\n")
display(data.nunique())

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY
1,-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY
2,-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY
3,-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY
4,-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY


-----Data Shape----

(20640, 10)

-----Description-----



,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
count,20640.000000,20640.000000,20640.000000,20640.000000,20433.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,-119.569704,35.631861,28.639486,2635.763081,537.870553,1425.476744,499.539680,3.870671,206855.816909
std,2.003532,2.135952,12.585558,2181.615252,421.385070,1132.462122,382.329753,1.899822,115395.615874
min,-124.350000,32.540000,1.000000,2.000000,1.000000,3.000000,1.000000,0.499900,14999.000000
25%,-121.800000,33.930000,18.000000,1447.750000,296.000000,787.000000,280.000000,2.563400,119600.000000
50%,-118.490000,34.260000,29.000000,2127.000000,435.000000,1166.000000,409.000000,3.534800,179700.000000
75%,-118.010000,37.710000,37.000000,3148.000000,647.000000,1725.000000,605.000000,4.743250,264725.000000
max,-114.310000,41.950000,52.000000,39320.000000,6445.000000,35682.000000,6082.000000,15.000100,500001.000000



----Information-----

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


None


-----Columns Name------



,Columns Name
0,longitude
1,latitude
2,housing_median_age
3,total_rooms
4,total_bedrooms
5,population
6,households
7,median_income
8,median_house_value
9,ocean_proximity



-----Before Missing Value------



longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64


-----After Missing Value------



longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64


-----Duplicated Value-----



np.int64(0)


------nunique Value-------



longitude               844
latitude                862
housing_median_age       52
total_rooms            5926
total_bedrooms         1923
population             3888
households             1815
median_income         12928
median_house_value     3842
ocean_proximity           5
dtype: int64

# Transform train & test separately

In [3]:
y = data["median_house_value"]
X = data.drop("median_house_value", axis=1)

X = pd.get_dummies(X, columns=["ocean_proximity"], drop_first=True)

In [4]:
# Train Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
#  Feature Scaling (Linear Model)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [5]:
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "Random Forest": RandomForestRegressor(random_state=42),
    "XGBoost": XGBRegressor(random_state=42)
}

results = []

for name, model in models.items():
    
    # Linear regression uses scaled data
    if name == "Linear Regression":
        model.fit(X_train_scaled, y_train)
        preds = model.predict(X_test_scaled)
        cv = cross_val_score(model, X_train_scaled, y_train, cv=5, scoring='r2')
    else:
        model.fit(X_train, y_train)
        preds = model.predict(X_test)
        cv = cross_val_score(model, X_train, y_train, cv=5, scoring='r2')
    
    r2 = r2_score(y_test, preds)
    rmse = np.sqrt(mean_squared_error(y_test, preds))
    
    results.append([name, r2, rmse, cv.mean()])


In [6]:
results_df = pd.DataFrame(results, columns=["Model", "R2 Score", "RMSE", "CV Mean R2"])
print(results_df.sort_values(by="R2 Score", ascending=False))

               Model  R2 Score          RMSE  CV Mean R2
3            XGBoost  0.829202  47309.128542    0.825657
2      Random Forest  0.816489  49038.209338    0.817913
1      Decision Tree  0.628676  69755.779954    0.637966
0  Linear Regression  0.625424  70060.521845    0.647542


# Conclusion

In this project, I worked on the California Housing dataset to predict median house prices using different machine learning models. I performed data cleaning, handled missing values, and applied for categorical features.

I trained and compared multiple models including Linear Regression, Decision Tree, Random Forest, and XGBoost using R² score, RMSE, and cross-validation.

# Final Result:

- XGBoost performed best (~0.83 R²)
- Random Forest also gave strong results
- Linear models performed lower due to non-linear data patterns

## Key Learning:

Ensemble models like Random Forest and XGBoost work better for real-world complex datasets compared to simple linear models